dartagnan/src/main/java/com/dat3m/dartagnan/program/event/CondJump.java

```diff
@@ -47,8 +54,26 @@ protected CondJump(CondJump other) {
 		notifier.addListener(this);
     }
 
+    private void initCtx() {
+        try {
+            Configuration config = Configuration.builder()
+            		.setOption("solver.z3.usePhantomReferences", "true")
+            		.build();
+			defaultCtx = SolverContextFactory.createSolverContext(
+			        config, 
+			        BasicLogManager.create(config), 
+			        ShutdownManager.create().getNotifier(), 
+			        Solvers.Z3);
+		} catch (Exception e) {
+			e.printStackTrace();
+		}
+    }
+    
```

Comment:
The default context should be created once globally. It should probably use the Solver specified by the User instead of Z3.

Reply:
I agree we can use a global context (maybe at the program level an initialise it when the program is initialise).
I think the used solver is irrelevant. This context is only used to check if a `CondJump` is a goto, i.e. its condition is always true.

Reply:
Initializing at the program level is fine I suppose. The used solver is irrelevant as long as all solvers are available. But if a user decides to e.g. only install PRINCESS on the system, then using Z3 will break. It is somewhat mitigated by the fact that Z3 is available for all operating systems but I still can imagine use-cases where Z3 might not be available for some reason or another.

Reply:
The problem that we have right now it that the information about which solver the user selected is not available in this class (or `Program.class`) and making that information available would require passing information which is irrelevant for anything else between classes.
Once we have a better way of handling configuration options as #79, this could be done much easier.

Reply:
Can't this information be part of the VerificationTask? Maybe in the Settings?

Reply:
It might, but I don't see the point in adding stuff to the Settings if we will (hopefully soon) get rid of this class

Reply:
I moved the `defaultCtx` to `Program` and "initialise" the field when the program is initialised.
The good thing about this is that when the program is initialised, the context makes use of the correct solver.
The problem is that there are many parts of the code (e.g. preprocessing, equivalence) that make use of `CondJump.isGoto()` (which requires the `defaultCtx`) and are called before `Program.initialise()`. I avoid a `NullExpception` problem by checking if the context was already initialised (which in most cases was not) and if not, I initialise it. Unfortunately here we run into the same problem that we cannot pass the correct solver so we use Z3 by default.
Your proposal of using the `VerificationTask` would have the same problem (unless we initialise the `defaultCtx` in one of the first passes of the program, i.e. before preprocessing and commuting equivalence classes).

Reply:
For now I'll revert this change since it results in the following problem with Z3
```
ASSERTION VIOLATION
File: ../src/ast/shared_occs.cpp
Line: 79
UNEXPECTED CODE WAS REACHED.
Z3 4.8.10.0
Please file an issue with this message and more detail about how you encountered it at https://github.com/Z3Prover/z3/issues/new
```
I suspect it is related to Z3Prover/z3#5107, but until JavaSMT does not release new bindings for 4.8.11.0 I cannot be sure.

Reply:
We can keep Z3 for now, but if you decide to do initialization this way, I would rather use a private static field and a public static getter. The getter can initialize everything if the value is still null (so it will automatically get initialized on first usage). But generally I would probably move the default context to a different place since Program is not the only class that could use a default context. If you have this new Helper class, maybe you can create a static method there that has the functionality that is needed by CondJump.isGoto(). Then the helper class would own the default context. But I suppose we can fix this problem outside of this PR.